### ISSUE TITLE
Feature request – gevent compatibility

### ISSUE THREAD

**Issue by davidkhess** [gorakhargosh/watchdog#306]
It would nice if watchdog was gevent compatible.

Problems I've run into so far:

* There is an os.read() in notify_c.py that is blocking – would need to use gevent.os.tp_read() instead.
* There's an indication that DelayedQueue is not working right. Not sure what is wrong there.

**Comment by fuag15** [comment]
this would be nice to have, i'm still digging into it but this is blocking me :+1:

**Comment by davidkhess** [comment]
Nice to have some company. :-)

FYI, my use case is trying to use Flask reloader. If you have watchdog installed the latest Flask automatically tries to use it – and if you are using gevent, it promptly locks up.

**Comment by davidkhess** [comment]
BTW, if anybody is looking for a strategy on how to support gevent, psycopg2's approach is pretty good:

http://initd.org/psycopg/docs/advanced.html#support-for-coroutine-libraries

Basically, if a callback gets registered, then the framework switches to non-blocking I/O and passes the responsibility for waiting on the I/O to the callback.

Psycogreen is an example of a module that you use with psycopg2 that provides the necessary callback to make gevent happy.

https://bitbucket.org/dvarrazzo/psycogreen

**Comment by freeauto** [comment]
+1

**Comment by gabrielgrant** [comment]
+1

**Comment by csantosbh** [comment]
Not sure you've had the same issue as me, but this seems to have fixed it for me: e4eebae522e2e6fc9448de12a8a15e924ed684e9

**Comment by Bogdanp** [comment]
I've run into this issue and created [watchdog_gevent](https://github.com/Bogdanp/watchdog_gevent). Hopefully it's of use to some of you.

**Comment by davidkhess** [comment]
@Bogdanp nice work. Is that a drop in replacement so that other tools that use watchdog work with it transparently? Or do they have to be coded to specifically use your package?

**Comment by Bogdanp** [comment]
@davidkhess It's drop-in. You `from watchdog_gevent import Observer` and it will return the best observer for the current platform. If gevent is available and `threading` is monkeypatched then it will return a `GeventObserver`, otherwise it will fall back to the observer returned by `watchdog` itself. Here's how I ended up using it in one of my projects:

https://github.com/Bogdanp/dramatiq/commit/ebf6e877f5e61bc6a707af575e7c1a95f35e3be7

**Comment by davidkhess** [comment]
The problem I ran into was the Flask reloader. Does Flask have to be modified to benefit from your package?

**Comment by Bogdanp** [comment]
Ah, I misunderstood. Yes, it probably does.

**Comment by BoboTiG** [comment]
What is the state of that issue watchdog 0.10.1?

**Comment by thomdask** [comment]
+1

**Comment by ddorian** [comment]
@davidkhess it seems to work from a quick test (ubuntu 22.04, flask 3.1.1, gevent 24.11.1, watchdog 6.0.0)

When I use `watchdog` there are 11 extra threads alive with `thread._target=None`:
```python
# put this inside a flask request
import threading

# print(["active=", threading.active_count()])
for thread in threading.enumerate():
    print([thread.name, thread.is_alive(), thread._target])


['MainThread', True, None]
['Thread-1', True, None]
['Thread-4', True, None]
['Thread-3', True, None]
['Thread-6', True, None]
['Thread-5', True, None]
['Thread-8', True, None]
['Thread-7', True, None]
['Thread-10', True, None]
['Thread-9', True, None]
['Thread-14', True, None]
['Thread-13', True, None]
['Thread-2 (serve_forever)', True, <bound method BaseWSGIServer.serve_forever of <werkzeug.serving.ThreadedWSGIServer object at 0x7fa415b62510>>]
['Thread-15 (process_request_thread)', True, <bound method ThreadingMixIn.process_request_thread of <werkzeug.serving.ThreadedWSGIServer object at 0x7fa415b62510>>]
```
